include/hipSYCL/algorithms/algorithm.hpp

```diff
@@ -383,6 +383,36 @@ sycl::event replace_copy(sycl::queue &q, ForwardIt1 first, ForwardIt1 last,
       new_value, deps);
 }
 
+template <class BidirIt, class ForwardIt>
+sycl::event reverse_copy(sycl::queue &q, BidirIt first,
+                         BidirIt last, ForwardIt d_first,
+                         const std::vector<sycl::event> &deps = {}) {
+  if (first == last)
+    return sycl::event{};
+
+  auto size = std::distance(first, last);
+
+  using value_type1 = typename std::iterator_traits<BidirIt>::value_type;
+  using value_type2 = typename std::iterator_traits<ForwardIt>::value_type;
+
+  // ??
+  if (std::is_trivially_copyable_v<BidirIt> &&
+      std::is_same_v<value_type1, value_type2> &&
+      util::is_contiguous<BidirIt>() && util::is_contiguous<ForwardIt>() &&
+      detail::should_use_memcpy(q.get_device())) {
+    return q.memcpy(&(*d_first), &(*first), size * sizeof(value_type1), deps);
+  } else {
```

Comment:
I would like to modify this if-condition but I don't exactly understand its purpose.  For `reverse_copy` this condition will always be false since the two types are different (Bidirectional and Forward iterators).  It's also in `std::copy`.

Reply:
This condition is there for algorithms like `copy`, because on some backends, using `memcpy` is faster than a custom kernel.

So, if we can use a bitwise-memcpy (e.g. if the type is trivially copyable) and it makes sense for the backend, use `memcpy`.

For `reverse_copy`, you can remove the `if` branch entirely because there's no way a `memcpy` can do a reverse as well. So this optimization won't apply.

> this condition will always be false since the two types are different (Bidirectional and Forward iterators)

I don't think this is the case, because we're comparing the value types, not the iterator types. So if the iterators point to values of the same type, the condition could be true.